File: Multivariate/Midterm/Midterm_4.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from factor_analyzer import FactorAnalyzer

#데이터 불려오기 
df = pd.read_csv("D:/knou/Multivariate/data/favoritesujects.csv")
#print(df.head())

# subject 열 제외
subjects = df.drop(columns=["SUBJECT"], errors="ignore")

# 1. 유의한 인자 수와 그 인자들이 확보하는 정보의 양은 얼마인가?  
# 초기 인자 분석
fa = FactorAnalyzer(rotation=None)
fa.fit(subjects)

# 고윳값 구하기 
ev, v = fa.get_eigenvalues()
#print(ev)

# 고윳값 ≥ 1 인자 수
num_factors = np.sum(ev >= 1)
#print(f"\n유의한 인자 수 (고윳값 ≥ 1): {num_factors}")

# 설명력 
#print(f"\n 변수 설명력 : {fa.get_factor_variance()}")

# 2. 인자부하행렬을 구하고 varimax와 promax 방법을 이용하여 인자회전을 실시하고 결과를 비교하시오
fa_varimax = FactorAnalyzer(n_factors=2 ,rotation='varimax', method='principal')
fa_varimax.fit(subjects)

# 인자계수
# print("인자계수 : ", fa_varimax.loadings_)

# # 인자공통성
# print("인자공통성 : ", fa_varimax.get_communalities())

# # 인자고유분산 : t-분산
# print("인자고유분산 : t-분산 : ", fa_varimax.get_uniquenesses())

# # 인자분석
# print("인자분석 :", fa_varimax.get_factor_variance())

# 2. 인자부하행렬을 구하고 varimax와 promax 방법을 이용하여 인자회전을 실시하고 결과를 비교하시오
fa_promax = FactorAnalyzer(n_factors=2 ,rotation='promax', method='principal')
fa_promax.fit(subjects)

# # 인자계수
# print("인자계수 : ", fa_promax.loadings_)

# # 인자공통성
# print("인자공통성 : ", fa_promax.get_communalities())

# # 인자고유분산 : t-분산
# print("인자고유분산 : t-분산 : ", fa_promax.get_uniquenesses())

# # 인자분석
# print("인자분석 :", fa_promax.get_factor_variance())

# # 요인간 상관행렬
# print("요인간 상관행렬 :", fa_promax.structure_) 

# 3인자들에 적절한 이름 (Python 언어)

File: Multivariate/Midterm/Midterm_4.R
# 300명의 대학생에게 6과목에 대해 선호도를 조사한 결과 이다 
# 인자분석을 실행하고 물어 답하라 

# 데이터 불려오기
subjects <-read.csv("D:/knou/Multivariate/data/favoritesujects.csv")
head(subjects)
summary(subjects)

# 라이브러리 설치
install.packages('GPArotation')
library(psych)
library(GPArotation)

# 4-1) 유의한 인자 수와 그 인자들이 확보하는 정보의 양은 얼마인가?
sub_factor = principal(subjects, rotate='none')

# 유의한 인자수 고윳값이 1 이상인 인자수
sum(sub_factor$values >= 1)

# 각 주성분의 고윳값(정보량)
sub_factor$values

# 설명력 (정보 비율)
explained_var <- sub_factor$values / sum(sub_factor$values)
explained_var

# 누적 설명력
cumulative_var <- cumsum(explained_var)
cumulative_var

# 4-2) 인자부하행렬을 구하고 varimax와 promax 방법을 이용하여 인자회전을 실시 하고 결과를 비교하시오

# 주성분 분석 (Varimax 회전)
sub_varimax <- principal(subjects, nfactors = 2, rotate = "varimax", score = TRUE, method = "regression")
sub_varimax

#promax
sub_varimax
principal(subjects, nfactor=2, rotate='promax', score=T, method="regression")


# 4-3) 인자들에 적절한 이름?

# 4-4) 인자분석 결과를 종합적으로 정리하시오